src/test/java/org/opentripplanner/gtfs/GtfsContextBuilder.java

```diff
@@ -0,0 +1,244 @@
+    }
+
+    public GtfsContextBuilder(GtfsFeedId feedId, OtpTransitServiceBuilder transitBuilder) {
+        this.feedId = feedId;
+        this.transitBuilder = transitBuilder;
+    }
+
+    public GtfsFeedId getFeedId() {
+        return feedId;
+    }
+
+    public OtpTransitServiceBuilder getTransitBuilder() {
+        return transitBuilder;
+    }
+
+    public GtfsContextBuilder withGraphBuilderAnnotationsAndDeduplicator(Graph graph) {
+        return withAddBuilderAnnotation(graph)
+                .withDeduplicator(graph.deduplicator);
+    }
+
+    public GtfsContextBuilder withAddBuilderAnnotation(AddBuilderAnnotation addBuilderAnnotation) {
+        this.addBuilderAnnotation = addBuilderAnnotation;
+        return this;
+    }
+
+    public GtfsContextBuilder withDeduplicator(Deduplicator deduplicator) {
+        this.deduplicator = deduplicator;
+        return this;
+    }
+
+    /**
+     * The {@link org.opentripplanner.graph_builder.module.GtfsModule} is responsible for repairing
+     * StopTimes for all trips and trip patterns generation, so turn this feature <b>off</b>
+     * when using GtfsModule to load data.
+     *
+     * This feature is turned <b>on</b> by <em>default</em>.
+     */
+    public GtfsContextBuilder turnOffRepairStopTimesAndTripPatternsGeneration() {
+        this.repairStopTimesAndGenerateTripPatterns = false;
+        return this;
+    }
+
+    public GtfsContextBuilder turnOnSetAgencyToFeedIdForAllElements() {
+        this.setAgencyToFeedIdForAllElements = true;
+        return this;
+    }
+
+    /**
+     * This method will:
+     * <ol>
+     *     <li>repair stop-times (if enabled)</li>
+     *     <li>generate TripPatterns (if enabled)</li>
+     *     <li>create a new context</li>
+     * </ol>
+     */
+    public GtfsContext build() {
+        if(repairStopTimesAndGenerateTripPatterns) {
+            repairStopTimesAndGenerateTripPatterns();
+        }
+        if (setAgencyToFeedIdForAllElements) {
+            setAgencyToFeedIdForAllElements();
+        }
+        return new GtfsContextImpl(feedId, transitBuilder);
+    }
+
+    /**
+     * By default this method is part of the {@link #build()} method.
+     * But in cases where you want to change the dao after building the
+     * context, and these changes will affect the TripPatterns generation,
+     * you should do the following:
+     *
+     * <pre>
+     * GtfsContextBuilder contextBuilder = &lt;create context builder>;
+     *
+     * // turn off TripPatterns generation before building
+     * context = contextBuilder
+     *     .turnOffRepairStopTimesAndTripPatternsGeneration()
+     *     .build();
+     *
+     * // Do you changes
+     * applyChanges(context.getDao());
+     *
+     * // Repair StopTimes and generate TripPatterns
+     * contextBuilder.repairStopTimesAndGenerateTripPatterns();
+     * </pre>
+     */
+    public void repairStopTimesAndGenerateTripPatterns() {
```

Comment:
I like that the stop time repair is factored out and made optional. However, it feels strange that this is done in GtfsContextBuilder, and that this class is reaching into an unfinished builder to carry out all these modifications at the moment of its build() operation. I see how it has evolved from the existing code. But again I hope we can eventually further refactor this to have more conventional behavior.